File: packages/hono/tests/email-start-hook.test.ts
import { describe, it, expect } from "vitest";
import { Hono } from "hono";
import Database from "better-sqlite3";
import { createAuth, runMigrations, defaultDeps, type EmailStart } from "@mattsmith/passkey-sdk-core";
import { mountAuthRoutes } from "../src/index.js";

describe("host email start on the mounted route", () => {
  for (const mode of ["success", "error", "empty", "whitespace", "invalid"] as const) {
    it(mode, async () => {
      const db = new Database(":memory:");
      try {
        runMigrations(db);
        let calls = 0, sends = 0;
        const start: EmailStart = async (input) => {
          calls++;
          expect(input.email).toBe("user@example.com");
          expect(input.expirySeconds).toBe(900);
          expect(input.now()).toBe(1700000000);
          expect(input.request?.headers.get("x-forwarded-for")).toBe("untrusted");
          if (mode === "error") throw new Error("synthetic reservation failure");
          return { otpId: mode === "empty" ? "" : mode === "whitespace" ? "  " : "opaque-host-id" };
        };
        const auth = createAuth({
          rpId: "example.com", origins: ["https://example.com"],
          session: { lifetimeSeconds: 3600, cookieName: "session" }, otp: { expirySeconds: 900 },
          email: { start, sendOtp: async () => { sends++; } },
          users: { findOrCreateByEmail: async () => { throw new Error("unexpected resolver"); } },
        }, { db, deps: { ...defaultDeps, now: () => 1700000000 } });
        const app = new Hono(); mountAuthRoutes(app, auth);
        const res = await app.request("/auth/email/start", {
          method: "POST", headers: { "content-type": "application/json", "x-forwarded-for": "untrusted" },
          body: JSON.stringify({ email: mode === "invalid" ? "not-an-email" : " USER@EXAMPLE.COM " }),
        });
        expect(res.status).toBe(mode === "success" ? 200 : mode === "invalid" ? 400 : 500);
        expect(calls).toBe(mode === "invalid" ? 0 : 1);
        expect(sends).toBe(0);
        expect(res.headers.get("set-cookie")).toBeNull();
        expect(db.prepare("SELECT COUNT(*) AS n FROM auth_email_otps").get()).toEqual({ n: 0 });
        if (mode === "success") expect(await res.json()).toEqual({ otpId: "opaque-host-id", expiresInSeconds: 900 });
      } finally { db.close(); }
    });
  }
});

File: servers/go/httpapi/email_start_hook_test.go
package httpapi

import (
	"context"
	"errors"
	"net/http/httptest"
	"strings"
	"testing"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/mattydsmith/passkey/servers/go/storage"
)

type startStorageSpy struct {
	storage.Storage
	writes int
}

func (s *startStorageSpy) CreateOTP(storage.EmailOTP) error { s.writes++; return nil }

type startSenderSpy struct{ calls int }

func (s *startSenderSpy) SendOTP(string, string) error { s.calls++; return nil }

func TestEmailStartHookMounted(t *testing.T) {
	for _, mode := range []string{"success", "error", "empty", "whitespace", "invalid"} {
		t.Run(mode, func(t *testing.T) {
			st, sender := &startStorageSpy{}, &startSenderSpy{}
			calls := 0
			now := time.Unix(1700000000, 0)
			key := struct{}{}
			cfg := Config{Storage: st, EmailSender: sender, RPID: "example.com", Origins: []string{"https://example.com"}, SessionCookieName: "session", OTPTTL: 15 * time.Minute, Now: func() time.Time { return now },
				EmailStart: func(ctx context.Context, in EmailStartInput) (EmailStartResult, error) {
					calls++
					if in.Email != "user@example.com" || in.OTPTTL != 15*time.Minute || !in.Now().Equal(now) || ctx.Value(key) != "request" || in.Request.RemoteAddr != "192.0.2.11:1234" || in.Request.Header.Get("X-Forwarded-For") != "untrusted" {
						t.Fatal("incorrect hook input")
					}
					if mode == "error" {
						return EmailStartResult{}, errors.New("synthetic reservation failure")
					}
					if mode == "empty" {
						return EmailStartResult{}, nil
					}
					if mode == "whitespace" {
						return EmailStartResult{OTPID: "  "}, nil
					}
					return EmailStartResult{OTPID: "opaque-host-id"}, nil
				},
			}
			r := chi.NewRouter()
			if err := Mount(r, cfg); err != nil {
				t.Fatal(err)
			}
			email := " USER@EXAMPLE.COM "
			if mode == "invalid" {
				email = "not-an-email"
			}
			req := httptest.NewRequest("POST", "/auth/email/start", strings.NewReader(`{"email":"`+email+`"}`))
			req = req.WithContext(context.WithValue(req.Context(), key, "request"))
			req.RemoteAddr = "192.0.2.11:1234"
			req.Header.Set("X-Forwarded-For", "untrusted")
			res := httptest.NewRecorder()
			r.ServeHTTP(res, req)
			want, callsWant := 500, 1
			if mode == "success" {
				want = 200
			}
			if mode == "invalid" {
				want = 400
				callsWant = 0
			}
			if res.Code != want || calls != callsWant || st.writes != 0 || sender.calls != 0 || len(res.Result().Cookies()) != 0 {
				t.Fatalf("status=%d hook=%d legacy writes=%d sends=%d", res.Code, calls, st.writes, sender.calls)
			}
			if mode == "success" && !strings.Contains(res.Body.String(), `"otpId":"opaque-host-id","expiresInSeconds":900`) {
				t.Fatalf("body=%s", res.Body.String())
			}
		})
	}
}
